### PR TITLE
Type-check object literal declaration

### PR DESCRIPTION
**Pull request by helixbass**
In this PR:
- type-check basic object literal variable declaration against annotated interface type

To test:
If you run `cargo run tmp.tsx` against a source file containing eg:
```
interface Foo {
  bar: number
}

const a: Foo = {
  bar: true,
}
```
you should see a diagnostic with `message_text: "Type 'boolean' is not assignable to type 'number'."` debug-logged
If you run `cargo run tmp.tsx` against a source file containing eg:
```
interface Foo {
  bar: number
}

const a: Foo = {
  bar: 1,
}
```
you shouldn't see any diagnostics debug-logged

### CODE REVIEW

File: src/compiler/types.rs [hunk 1]
```diff
@@ -1097,33 +1208,111 @@ impl Type {
...
             Type::IntrinsicType(intrinsic_type) => intrinsic_type.flags(),
             Type::LiteralType(literal_type) => literal_type.flags(),
+            Type::ObjectType(object_type) => object_type.flags(),
             Type::UnionOrIntersectionType(union_or_intersection_type) => {
                 union_or_intersection_type.flags()
             }
         }
     }
+
+    fn maybe_symbol(&self) -> Option<Rc<Symbol>> {
```
Comment (helixbass) [inline src/compiler/types.rs:1244]: It's probably time to implement a similar boilerplate-generation macro for type types as I did for node types

File: src/compiler/checker.rs [hunk 2]
```diff
@@ -1026,12 +2465,30 @@ impl NodeBuilder {
...
+        &self,
+        tracker: Option<&dyn SymbolTracker>,
+        cb: TCallback,
     ) -> Option<TReturn> {
-        let context = NodeBuilderContext::new(tracker.unwrap());
+        let default_tracker: Option<DefaultNodeBuilderContextSymbolTracker> = if tracker.is_some() {
+            None
+        } else {
+            Some(DefaultNodeBuilderContextSymbolTracker::new())
+        };
```
Comment (peterstuart) [inline src/compiler/checker.rs:2489]: A `match` seems like the more idiomatic thing to do here, to me.

**Review by peterstuart — COMMENTED** [review]
This is more than I can really wrap my head around, so I skimmed some of it, but left a few comments.

File: src/compiler/checker.rs (hunk 2 above)
Comment (helixbass) [reply src/compiler/checker.rs:2488]: Ok updated to use `match`

File: src/compiler/binder.rs [hunk 3]
```diff
@@ -162,7 +162,7 @@ impl BinderType {
         ) {
             let mut members = symbol.maybe_members();
             if members.is_none() {
-                *members = Some(create_symbol_table());
+                *members = Some(Rc::new(RefCell::new(create_symbol_table())));
```
[the remark this answers is not in the document]
Comment (helixbass) [reply src/compiler/binder.rs:165]: That's an interesting idea, but seems like then during creation (before I've wrapped in an `Rc`/`RefCell`) I'd need "temporary" "non-wrapped" versions of the types?

[the remark this answers is not in the document]
Comment (peterstuart) [reply src/compiler/binder.rs:165]: Right... I guess the macro could generate some private `BlahInternal` type based on the struct `Blah`? Seems pretty complicated.

**Pull request merged**